### PR TITLE
executor: grant statement allow pattern string in db and table name

### PR DESCRIPTION
**Pull request by tiancaiamao**
We check the database name very strict before, but maybe user want to specific pattern match for grant statement.

For https://github.com/pingcap/tidb/issues/2456

### CODE REVIEW

File: executor/grant.go [hunk 1]
```diff
@@ -246,30 +248,31 @@ func (e *GrantExec) grantDBPriv(priv *ast.PrivElem, user *ast.UserSpec) error {
...
 	}
+	tblName := e.Level.TableName
 	userName, host := parseUser(user.User)
-	asgns, err := composeTablePrivUpdate(e.ctx, priv.Priv, userName, host, db.Name.O, tbl.Meta().Name.O)
+	asgns, err := composeTablePrivUpdate(e.ctx, priv.Priv, userName, host, dbName, tblName)
 	if err != nil {
 		return errors.Trace(err)
 	}
-	sql := fmt.Sprintf(`UPDATE %s.%s SET %s WHERE User="%s" AND Host="%s" AND DB="%s" AND Table_name="%s";`, mysql.SystemDB, mysql.TablePrivTable, asgns, userName, host, db.Name.O, tbl.Meta().Name.O)
+	sql := fmt.Sprintf(`UPDATE %s.%s SET %s WHERE User="%s" AND Host="%s" AND DB="%s" AND Table_name="%s";`, mysql.SystemDB, mysql.TablePrivTable, asgns, userName, host, dbName, tblName)
```
Comment (hanfei1991) [inline executor/grant.go]: change `Table_name =` to `Table_name like` ?

Comment (tiancaiamao) [inline executor/grant.go]: select use `like`, but insert and update should use `=`
update privilege just update one item.

Comment (hanfei1991) [inline executor/grant.go]: But it seems you don't fetch all possible dbs and tables.

Comment (tiancaiamao) [inline executor/grant.go]: When I want to grant privilege of all database that begin with name 'te%' to someone, 
it doesn't insert multiple records, but just one, into mysql.db, like this:

```
+------+-----+--------+-------------+-------------+-------------+-------------+-------------+-----------+------------+------------+------------+--------------+
| Host | DB  | User   | Select_priv | Insert_priv | Update_priv | Delete_priv | Create_priv | Drop_priv | Grant_priv | Index_priv | Alter_priv | Execute_priv |
+------+-----+--------+-------------+-------------+-------------+-------------+-------------+-----------+------------+------------+------------+--------------+
| %    | te% | genius | Y           | Y           | Y           | Y           | Y           | Y         | Y          | Y          | Y          | Y            |
+------+-----+--------+-------------+-------------+-------------+-------------+-------------+-----------+------------+------------+------------+--------------+
``` 

In the grant logic, this kind of record will be checked, if not exist, create one first.
Update will update this record, and use exactly `=`.

When used for privilege check, this record will be Select use `like` pattern match, 
so we know all database that begin with name 'te%' apply the rule of privilege in the record.

Comment (hanfei1991) [inline executor/grant.go]: Well, It's too strange.

**Comment by zimulala** [comment]
Fix conflicts.

**Comment by tiancaiamao** [comment]
DONE @zimulala

**Comment by zimulala** [comment]
LGTM

**Pull request merged**